Single implementation of a systema conn pool

josecv opened the pull request:
Signed-off-by: Jose Cortes <josecortes@datawire.io>

## Description

This'll make it so there's a single implementation of the systema conn pool with uniform headers passed through on all requests. Part of some lifting so we can pass the install id to systema requests

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

josecv committed:
what would it take to use go 1.18?

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Pull out a generic way to construct systema connection pools

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Replace user daemon systema connections with a8rcloud pool

Signed-off-by: Jose Cortes <josecortes@datawire.io>

thallgren commented on cmd/traffic/cmd/manager/manager.go:
```diff
@@ -136,9 +137,13 @@ func (m *Manager) runSystemAGCLoop(ctx context.Context) error {
+					if err := systema.Done(ctx); err != nil {
 						dlog.Errorln(ctx, "systema: release management connection:", err)
 					}
+					// Double free needed to decrease refcount of connection acquired, but not released, in UpdateIntercept
```
Where's the call to `UpdateIntercept`? And why isn't `UpdateIntercept` releasing the connection that it acquired?

josecv replied:
So it's not a call that comes from here, but rather the original UpdateIntercept call that created the domain --it [acquires a connection](https://github.com/telepresenceio/telepresence/blob/release/v2/cmd/traffic/cmd/manager/service.go#L466) but doesn't drop it except [for](https://github.com/telepresenceio/telepresence/blob/release/v2/cmd/traffic/cmd/manager/service.go#L508-L510) [errors](https://github.com/telepresenceio/telepresence/blob/release/v2/cmd/traffic/cmd/manager/service.go#L541-L543). I dug through the history and this seems to be deliberate: the pool's .Get() sets up a reverse connection to systema, so that has to stay open as long as there's an intercept with a preview url live

thallgren replied:
So the `update` that is read from the `snapshot.Updates` "carries" an open connection with it (the comment could be clearer about that)? Isn't there a risk for a leak here? What if the `systema.Get` fails on line 122 for instance? Shouldn't the second `system.Done` be called regardless?

josecv replied:
> So the update that is read from the snapshot.Updates "carries" an open connection with it (the comment could be clearer about that)?

yes it basically does --for it to get here that connection must've been opened by an updateintercept rpc

> Isn't there a risk for a leak here? What if the systema.Get fails on line 122 for instance? Shouldn't the second system.Done be called regardless?

yes great point, i'll go change this

thallgren commented on pkg/a8rcloud/systema.go:
```diff
@@ -0,0 +1,169 @@
+
+type SystemAPool[T Closeable] interface {
+	Get(ctx context.Context) (T, error)
+	Done(ctx context.Context) error
```
An alternative to a Get/Done pair could be:
```
   Do(context.Context, func(context.Context, T) error) error
```
that would use a private get/done pair internally.

josecv replied:
I'd love to do that, the only issue is the way UpdateIntercept currently expects the connection to be left dangling --it'd have to be changed to have a long-running goroutine that mirrors the intercept lifecycle

thallgren replied:
Yeah, I understand that now. Never mind.

josecv committed:
Fix linter for go 1.18

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
make generate

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Address PR comment

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Fix a generic cast.

Turns out generics of interfaces aren't the same as the generic of the
implementation

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
bump helm chart

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Use the right context for long-lived systema connections

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
Add the UserDaemonSystemAProxyClient to the session clients

Signed-off-by: Jose Cortes <josecortes@datawire.io>

josecv committed:
bump circle to go 1.18

Signed-off-by: Jose Cortes <josecortes@datawire.io>

Commit:
Fix instability issue in TestGatherLogs_OnlyMappedLogs

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv committed:
Generate certs with install test

Signed-off-by: Jose Cortes <josecortes@datawire.io>
